### PR TITLE
ar71xx: add support for TP-Link RE450

### PR DESCRIPTION
**Pull request by dasty**
This merge request is based on previous patches submitted to OpenWRT mailing list as mentioned in the commit messages. It has been tested on my TP-RE450 and it seems to be running nicely for last couple of days without any issue.

Additionally one user/contributor reported success with earlier version of the patches:
https://forum.openwrt.org/viewtopic.php?pid=343055#p343055

NOTE: Question of merging "kernel image generator for TP-Link RE450" with some other tools, have been discussed and it does not seem to make any sense: https://lists.openwrt.org/pipermail/openwrt-devel/2016-January/039048.html

After re-basing to latest master I just re-compiled the patches and everything seems to be fine. Extensive testing was done based on cc5c8f681e05f18678c8b10b9f747689e875ceaf

### CODE REVIEW

**Comment by mkresin** [comment]
@jow- You have the same commit in your staging tree. How to deal with it?

**Comment by dasty** [comment]
I took a short look at https://git.lede-project.org/?p=lede/jow/staging.git;a=commit;h=645536aeca04bafa49bf0f9b2c31dc74c1e2f0c0 and that commit seems to be incomplete. It is the code I based my work on top of, but some things are missing. If that commit goes in, I would have to rebase my changes on top of it and make it work again. For sure statement "The result factory image is flashable from the stock web ui." in that commit is not valid.

**Comment by jow-** [comment]
@mkresin - I started refactoring an older pull request which was inadequate for inclusion but did never finish it. Please prefer @dasty's variant if it is complete - I'll happily rebase my tree on top of it then.

**Comment by mkresin** [comment]
@jow- is this one now adequate for inclusion? I do **not** plan to pick this. I've way to less knowledge about ar71xx.

@dasty would you please rebase your PR on LEDE Head. Due to commits to LEDE there is a conflicting file now.

**Comment by dasty** [comment]
sure I will rebase. Due to fd924d2068f2 it is not just trivial rebase. Anything else what should be fixed?

File: target/linux/ar71xx/base-files/etc/hotplug.d/firmware/11-ath10k-caldata [hunk 1]
```diff
@@ -94,6 +94,10 @@ case "$FIRMWARE" in
 	unifiac-pro)
 		ath10kcal_extract "EEPROM" 20480 2116
 		;;
+	re450)
```
Comment (mkresin) [inline target/linux/ar71xx/base-files/etc/hotplug.d/firmware/11-ath10k-caldata]: please keep alphabetical order

**Comment by dasty** [comment]
I just pushed PATCHv2. Changes compare to v1 are:

1) additional commit sorting 11-ath10k-caldata alphabetically, it can be merged without this commit as well.

2) rebased on top of latest master - involved replacing function do_re450 with a struct e450_info in tplink-safeloader.c

everything nicely compiles and build images seems to be valid.

**Comment by dasty** [comment]
I just rebased it to the master again, compiled and flashed it to my device. Seems to be running quite fine.

@jow- Do I assume correctly it will be you, who will be doing the merge or at least taking it into your staging tree? If yes, could you please provide an indication, when you will have some time to take a look, so I can make sure there are no merging conflicts.

**Comment by dasty** [comment]
Hi everybody, is there anything I can do to get this patches merged. Should I fix, rewrite something? @jow- should I rebase on top of your staging tree? Would that make your life easier?

**Comment by jow-** [comment]
@dasty - if you could do that I would be most grateful. I have to admit that this PR fell out of my attention span :(

I'm especially interested if the safeloader changes in my staging tree are working properly for you.

**Comment by dasty** [comment]
@jow- I really liked your changes to tplink-safeloader.c and they are working nicely for me. I re-based on top of your master, bit split the commits, so they made sense and to avoid conflicts, recompiled and flashed using the sysupgrade and everything seems to be working fine. Additionally I checked factory images using binwalk and they seem to be very similar to previous ones.

Everything is pushed into https://github.com/dasty/lede-source/tree/TP-RE450-jow (I did not want to push to the old branch as it would make your staging tree part of this change request)

**Comment by jow-** [comment]
Thank you @dasty ! I integrated your branch into my staging tree and had to rebase some of my changes on top of the current LEDE master and then rebased your branch on top of mine.

I squashed your no-nand and LED rename fixups into the original commit so that proper support is introduced on the get-go when finally merging the changes.

**Comment by dasty** [comment]
@jow- I had a short look at your master branch and it looks good to me. Thanks for merging it to your staging and looking forward to seeing changes in the master eventually.

**Comment by alxscott** [comment]
Apologies to re-open this thread, but it seems best placed... I just bough 2 x TP-Link RE450 and it is hardware variant V2.. The LEDE image fails checksum check for install. Device Wiki ( https://wikidevi.com/wiki/TP-LINK_RE450) shows same CPU and Wireless chips, so I'm hoping it is a simple code tweak? 

Many Thanks

Alex

**Comment by dasty** [comment]
@alxscott - please read my answer on the forum https://forum.openwrt.org/viewtopic.php?pid=368144#p368144 - this is not a good place to discuss this.

**Pull request closed without merging**